File: EXAMPLES/LeetCode/0082-remove-duplicates-from-sorted-list-II-MEDIUM.js
// 2022-05-30
// https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var deleteDuplicates = function(head) {

    // fake head, will not move
    const start = new ListNode(0, head);

    // last node before duplicates sub list
    let prev = start;
    
    while (head) {
        if (head.next && head.val === head.next.val) {
            while (head.next && head.val === head.next.val) {
                head = head.next;
            }
            prev.next = head.next;
            head = head.next;
        } else {
            prev = prev.next;
            head = head.next;
        }
    }
    return start.next;
};

File: EXAMPLES/LeetCode/0082-remove-duplicates-from-sorted-list-II-MEDIUM.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:

        # fake head, will not move
        start = ListNode(0, head)
        
        # predecesor, will move
        pred = start

        while head:
            if head.next and head.val == head.next.val:
                while head.next and head.val == head.next.val:
                    head = head.next
                pred.next = head.next
                head = head.next
            else:
                pred = pred.next
                head = head.next
            
        return start.next
